vowpalwabbit/gd.h

```diff
@@ -65,23 +65,37 @@ inline void vec_add_multipredict(multipredict_info<T>& mp, const float fx, uint6
 template <class DataT, class WeightOrIndexT, void (*FuncT)(DataT&, float, WeightOrIndexT)>
 inline void foreach_feature(vw& all, example& ec, DataT& dat)
 {
-  return all.weights.sparse
-      ? foreach_feature<DataT, WeightOrIndexT, FuncT, sparse_parameters>(all.weights.sparse_weights,
-            all.ignore_some_linear, all.ignore_linear, *ec.interactions, all.permutations, ec, dat)
-      : foreach_feature<DataT, WeightOrIndexT, FuncT, dense_parameters>(all.weights.dense_weights,
+  if(all.weights.sparse)
+  {
+    all.weights.sparse_weights.set_tag(hashall(ec.tag.begin(),ec.tag.size(),all.hash_seed)%32);
+    return foreach_feature<DataT, WeightOrIndexT, FuncT, sparse_parameters>(all.weights.sparse_weights,
             all.ignore_some_linear, all.ignore_linear, *ec.interactions, all.permutations, ec, dat);
+  }
+  else
+  {
+    all.weights.dense_weights.set_tag(hashall(ec.tag.begin(),ec.tag.size(),all.hash_seed)%32);
```

Comment:
we should unset the tag before we return from here  (and the foreach just below)
i.e. 

```
sethash...
foreach_feature...
unsethash...
```

I think we can remove the return since the function signature return type is void